test/e2e/instrumentation/monitoring/prometheus.go

```diff
@@ -0,0 +1,389 @@
+	})
+	It("should scrape metrics from annotated services.", func() {
+		var customMetricTargetValue float64 = 100
+		var consumerName string = "prometheus-custom-service-metric"
+		var requiredSampleCount int = 1
+		var annotatePods, annotateService bool = false, true
+		query := prometheusServiceCustomMetricQuery(f.Namespace.Name, consumerName)
+		consumer := exportCustomMetric(f, consumerName, int(customMetricTargetValue), annotatePods, annotateService)
+		defer consumer.CleanUp()
+		retryUntil(func() error {
+			return validateQueryReturnsCorrectValues(f.ClientSet, query, customMetricTargetValue, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, prometheusTestTimeout)
+	})
+})
+
+func prometheusCPUQuery(namespace, podNamePrefix string, rate time.Duration) string {
+	return fmt.Sprintf(`sum(irate(container_cpu_usage_seconds_total{namespace="%v",pod_name=~"%v.*"}[%vm]))`,
+		namespace, podNamePrefix, int64(rate.Minutes()))
+}
+
+func prometheusServiceCustomMetricQuery(namespace, service string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%v",kubernetes_name="%v"})`, namespace, service)
+}
+
+func prometheusPodCustomMetricQuery(namespace, podNamePrefix string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%s",kubernetes_pod_name=~"%s.*"})`, namespace, podNamePrefix)
+}
+
+func consumeCPUResources(f *framework.Framework, consumerName string, cpuUsage int) *common.ResourceConsumer {
+	return common.NewDynamicResourceConsumer(consumerName, f.Namespace.Name, common.KindDeployment, 1, cpuUsage,
+		memoryUsed, 0, int64(cpuUsage), memoryLimit, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func exportCustomMetric(f *framework.Framework, consumerName string, metricValue int, annotatePods, annotateService bool) *common.ResourceConsumer {
+	var podAnnotations, serviceAnnotations map[string]string
+	if annotatePods {
+		podAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	if annotateService {
+		serviceAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	return common.NewMetricExporter(consumerName, f.Namespace.Name, podAnnotations, serviceAnnotations, metricValue, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func validateMetricAvailableForNodes(c clientset.Interface, metric string, expectedNodesNames []string) error {
+	instanceLabels, err := getInstanceLabelsAvailableForMetric(c, prometheusRequiredNodesUpDuration, metric)
+	if err != nil {
+		return err
+	}
+	nodesWithMetric := make(map[string]bool)
+	for _, instance := range instanceLabels {
+		nodesWithMetric[instance] = true
+	}
+
+	for _, nodeName := range expectedNodesNames {
+		if _, found := nodesWithMetric[nodeName]; !found {
+			return fmt.Errorf("Metric '%v' is not available for node '%v'", metric, nodeName)
```

Comment:
I think the number of nodes would be more useful, for automatically running tests the cluster is deleted before anyone can ssh to a node and debug it. And information if the metrics are available for all nodes vs most nodes vs no nodes already gives some idea.

Reply:
ok, "3 out of 5 nodes"